### DISCUSSION TITLE
Are Rotary Encoders being debounced?

### DISCUSSION THREAD

**Discussion by glyndon** [arendst/Tasmota#11221] [Q&A] [answered]
I have an ESP-based lamp dimmer running my own code, and it handled the encoder without ever 'skipping' or yeilding surprising results (once I added debounce logic, that is). 
So then I installed Tasmota 9.3.1 on that device, and once configured to use the encoder to set dimmer level it works as it should.
But the encoder now can produce sudden jumps in light level, sometimes opposite the way it's being turned.
This has every symptom of there being no debouncing timer on the encoder inputs.
I can't find anything definitive about ratery debouncing either way in the docs, but if it's not yet implemented for rotary's, can this be a feature request?

**Comment by barbudor** [comment, marked as answer]
There is reference to deboucing in tasmota\support_rotary.ino but seems to be using a fixed 10ms value which can be overridden in user_config_override with 
` #define ROTARY_DEBOUNCE      value-in-ms` 
Can't confirm if this is what would cause what you are describing (which I also witnessed).

What debounce time are you using in you own code ?
Can you test by recompiling with a higher value ?

**Reply by glyndon** [reply]
Helpful to see it's there. Perhaps just exposing it to the UI for reconfiguration would be helpful to others using possibly-flaky encoders. Or use a state machine.
In my code I usually write in a debounce timer, but now recall that in the device I'm describing I had instead used this library. This library uses a state machine instead of debouncing timers, so it can handle even rapid turning of the knob, which it does very well.
https://github.com/MajicDesigns/MD_REncoder
The underlying algorithm is well-documented and used in several popular rotary libraries.

**Comment by glyndon** [comment]
I compiled Tasmota after changing
#define ROTARY_DEBOUNCE
from 10 to 50.
No improvement in usability, it still sometimes jumps to random brightnesses while the rotary is turning.
It should have made *some* difference, but I could not see any.

I suggest replacing the debouncing code (which has a comment indicating it may be somewhat tentative),
with the state-machine logic used by the library I mentioned earlier. 
That algorithm works flawlessly on this very same hardware (aged noisy encoder and all).
